Add UI OfferBook list add/remove checks

ghubstan opened the pull request:
Problem:  Using the API's CLI to edit offers causes add/remove messages to be received on peers in the same envelope batch, then passed to the UI in random order.  When `OfferBook` receives a remove(offer) event after an add(offer) event, the edited offer will be removed immediately after being added to the UI's `OfferBook` list.  

Fix:  This change uses storage entry sequence number and storage entry payload hash comparisons to avoid the problem.

- `OfferBookListItem` Added new constructor taking `P2PDataStorage.ByteArray hashOfPayload`,
  and `int sequenceNumber` params.  Defined a `toString` method.
  
- `OfferBook` Added new checks on `OfferBookListItem`'s `hashOfPayload` and `sequenceNumber` while
  determining if offer candidates should be added or removed from the UI's OfferBook List.
  See OfferBook contructor's implementation of OfferBookChangedListener#onAdded and
  OfferBookChangedListener#onRemoved.  Added many (maybe too many) comments explaining the add/remove rules,
  and plenty of debug statements (maybe too many) to help trace the add/remove event process.
  
- `OfferBookService#OfferBookChangedListener` Added new `P2PDataStorage.ByteArray` hashOfPayload,
  and `int sequenceNumber` params to listener's `onAdded` and `onRemoved` method signatures.
  Added these two new parameter values to listener.onAdded and listener.onRemoved calls.
  
- `TakeOfferDataModel` Replaced unused `tradeManager` param in `offerBook.removeOffer`
  with `P2PDataStorage.ByteArray hashOfPayload` (null), and `int sequenceNumber` (-1) params.
  `OfferBook` will remove the candidate offer as before.

- MarketAlerts` Adjusted `onAdded` & `onRemoved` listener method signatures.

This PR is based on https://github.com/bisq-network/bisq/pull/5577, which should be reviewed and merged before this one.

ghubstan committed:
Refactor EditOfferDataModel for new editoffer api method

- Define set of editable offer payload fields in MutableOfferPayloadFields.

- Move bulk of EditOfferDataModel#onPublishOffer logic to OfferUtil.

ghubstan committed:
Fix comment

ghubstan committed:
Add OfferInfo field isActivated, rpc EditOffer to proto

ghubstan committed:
Add server/core editOffer, adjust getMyOffer(s) impls

- Add editOffer to GrpcOffersService, CoreApi, CoreOffersService.

- Set editOffer call rate meter to 1 / minute.

- Use new EditOfferValidator to verify editOffer params OK.

- Adust getMyOffer(s) rpc impl and OfferInfo model to use OpenOffer
  for accessing activation state and trigger price.

ghubstan committed:
Refactor GrpcClient: request builders moved bisq.cli.request pkg

Reduces size of GrpcClient while allowing for additional bot-friendly
variations of the new grpc editOffer method.

ghubstan committed:
Add new EditOfferOptionParser and test

ghubstan committed:
Add editoffer method help

ghubstan committed:
Add editoffer to CLI

ghubstan committed:
Add editoffer api tests & minor apitest refactoring

ghubstan committed:
Remove chase quickpay acct test

ghubstan committed:
Remove unused import

ghubstan committed:
Fix problems found in codacy check

ghubstan committed:
Fix codacy issue

ghubstan committed:
Fix log arg spec bug

ghubstan committed:
Add bool isMyOffer to OfferInfo proto

ghubstan committed:
Fix editoffer validation bugs, tidy up CoreOffersService

ghubstan committed:
Show enable/trigger-price cols for 'getmyoffer'

ghubstan committed:
Improve 'editoffer' opt parsing, fix test pkg name

ghubstan committed:
Add editoffer test case, suppress annoying warnings

ghubstan committed:
Remove unused field

ghubstan committed:
Throw exception is edit altcoin offer is attempted

Support for editing BSQ offers is in place, but will be added
in another PR.

ghubstan committed:
Avoid duplicate test run

ghubstan committed:
Make codacy just a bit happier

ghubstan committed:
Force codacy check after codacy config change

ghubstan committed:
Include isMyOffer flag in API's trade/offer proto wrappers

Optionally displaying an ENABLED column in CLI side getoffer output
depends on the value of offer.isMyOffer, which is passed via new
boolean arguments to the trade & offer pojo builders.

ghubstan committed:
Pass isMyOffer flag to trade/offer proto wrappers from core services

ghubstan committed:
Add altcoin (bsq) offer editing validation check

BSQ offers are fixed-price only.  This change blocks an
attempt to change an altcoin offer to a margin price based
offer, or set a trigger price.

ghubstan committed:
Optionally show ENABLED column in CLI's getoffer(bsq) output

ghubstan committed:
Add BSQ offer editing tests to EditOfferTest

And log CLI's getoffer output to see getoffer formatting -- after adding
new ENABLED and TRIGGER-PRICE columns.

ghubstan committed:
Merge branch 'master' into 02-refactor-for-api-editoffer-method

ghubstan committed:
Merge branch '02-refactor-for-api-editoffer-method' into 03-api-editoffer-method

ghubstan committed:
Merge branch '03-api-editoffer-method' into 04-edit-bsq-offer

ghubstan committed:
Force rebuild after github action ECONNRESET

ghubstan committed:
Document api 'editoffer' usage

ghubstan committed:
Fix header

ghubstan committed:
Delete tmp main() method

ghubstan committed:
Rename and move private function

And make sure function is not duplicated CLI side logic.

ghubstan committed:
Do not duplicate Price.parse on CLI side for only one use case

ghubstan committed:
Remove old TODO because relevant refactoring was approved

ghubstan committed:
Fix typo 'enabled' -> 'enable'

ghubstan committed:
Always use Locale.US in CLI DecimalFormats

Avoid inconsistent CLI output decimal formats across different
systems' default locales.

ghubstan committed:
Merge branch 'master' into 05-update-api-beta-test-guide

ghubstan committed:
Do not start test harness deamons in dbg mode by default

ghubstan committed:
Add missing trigger-price param

ghubstan committed:
Fix peer add(offer) & remove(offer) event order problem

Use LinkedHashSet to maintain NetworkEnvelope ordering when Connection#onBundleOfEnvelopes
calls listeners.

Connection#onBundleOfEnvelopes builds a set from an ordered list of NetworkEnvelopes,
then calls listeners during set iteration.  The envelope list ordering is lost if a
HashSet is built, but maintained by switching to a LinkedHashSet.

Losing the envelope ordering becomes a problem if the peer receives a RemoveDataMessage
and an AddDataMessage in the same batch of envelopes, and relays them to listeners
in the wrong (random) order.  For example, an API 'editoffer' call may result in the
edited offer being added, then immediately remove from their UI's offer book.

ghubstan committed:
Revert "Fix peer add(offer) & remove(offer) event order problem"

This reverts commit add6536402daa36330711d6b73ad99157e00187e.

ghubstan committed:
Merge branch 'master' into 05-update-api-beta-test-guide

ghubstan committed:
Ensure UI OfferBook items are correctly added and removed

Using the API's CLI to edit offers can sometimes result in add/remove messages
being received on peers in the same batch of envolopes, and these messages
are sometimes passed to the UI in (1) add, (2) remove order.  This can result in
a newly edited offer being removed immediately after being added to the OfferBook
list. This change uses storage entry sequence number and storage entry payload
hash comparisons to avoid the problem.

- OfferBookListItem Added new constructor taking P2PDataStorage.ByteArray hashOfPayload,
  and int sequenceNumber params.  Added a new toString() method.

- OfferBook Added new checks on OfferBookListItem hashOfPayload and sequenceNumber while
  determining if offer candidates should be added or removed from the UI's OfferBook List.
  See OfferBook contructor's implementation of OfferBookChangedListener#onAdded and
  OfferBookChangedListener#onRemoved.  Added many comments explaining the add/remove rules,
  and plenty of debug statements to help trace the add/remove event process.

- OfferBookService#OfferBookChangedListener Added new P2PDataStorage.ByteArray hashOfPayload,
  and int sequenceNumber params to listener's onAdded and onRemoved method signatures.
  Added these two new paramater values to listener.onAdded and listener.onRemoved calls.

- TakeOfferDataModel Replaced unused, old tradeManager param in offerBook.removeOffer()
  with (null) P2PDataStorage.ByteArray hashOfPayload, and (-1) int sequenceNumber params.
  OfferBook will remove the candidate offer as before.

- MarketAlerts Adjusted onAdded() & onRemoved listener method signatures, even though
  new P2PDataStorage.ByteArray hashOfPayload, int sequenceNumber params are not used
  by the implementations.

ghubstan committed:
Revert logback debug config changes

(Accidentally included in last commit.)

ghubstan committed:
Merge branch 'master' into 06-safe-offerbook-add-remove-events

ghubstan committed:
Remove comment (question answered)

sqrrm commented on desktop/src/main/java/bisq/desktop/main/offer/offerbook/OfferBook.java:
```diff
@@ -83,44 +85,138 @@ public void onAdded(Offer offer) {
-                                "old offerBookListItem={}, new offerBookListItem={}", candidateWithSameId.get(), offerBookListItem);
-                        offerBookListItems.remove(candidateWithSameId.get());
+                    OfferBookListItem newOfferBookListItem = new OfferBookListItem(offer, hashOfPayload, sequenceNumber);
+                    removeAnyOldOfferBookListItemsBeforeAddingReplacement(newOfferBookListItem);
```
I find this method name somewhat on the long side

ghubstan replied:
Requested change in commit 59c031327725bfc669414389056b875eff9a9faa.

sqrrm commented on desktop/src/main/java/bisq/desktop/main/offer/offerbook/OfferBookListItem.java:
```diff
@@ -79,7 +110,7 @@ public WitnessAgeData getWitnessAgeData(AccountAgeWitnessService accountAgeWitne
                     // Or banned
                     daysSinceSignedAsLong = TimeUnit.MILLISECONDS.toDays(optionalWitness.map(witness ->
-                            accountAgeWitnessService.getWitnessSignAge(witness, new Date()))
+                                    accountAgeWitnessService.getWitnessSignAge(witness, new Date()))
```
Bad indent

ghubstan replied:
Either, this is formatting of an older commit (which I have not found yet).  Or my IDE style rules have changed since I upgraded to latest version (doubtful because my bisq style rule file has not changed).

Lines 112-114 in the repo and after auto-formatting in my IDE are
```
                    daysSinceSignedAsLong = TimeUnit.MILLISECONDS.toDays(optionalWitness.map(witness ->
                                    accountAgeWitnessService.getWitnessSignAge(witness, new Date()))
                            .orElse(0L));

```

ghubstan committed:
Handle API's edit+disable offer use case in UI.

This change is a refactoring for handling the removal of a peer UI's offer
item when it is deactivated and edited in the same CLI `editoffer` command.

On the API side, an `editoffer --price=N --enable=false` command results
in the edited offer not being re-published.  On a UI peer's side, the edited
offer is not added to the peer's storage, and the peer's onRemoved(offer)
listener event does not find a storage entry with matching payload-hash.

This fix assumes an offer that is not in the local store should be removed
from the UI's view list -- when the onRemoved method's hashOfPayload does
not match the UI's view list item's hashOfPayload.

sqrrm commented on desktop/src/main/java/bisq/desktop/main/offer/offerbook/OfferBook.java:
```diff
@@ -229,8 +247,8 @@ public void fillOfferBookListItems() {
-                    .filter(o -> !filterManager.isOfferIdBanned(o.getId()))
-                    .filter(o -> !OfferRestrictions.requiresNodeAddressUpdate() || Utils.isV3Address(o.getMakerNodeAddress().getHostName()))
+                    .filter(o -> !isOfferIdBanned(o))
+                    .filter(o -> isV3NodeAddressCompliant(o))
```
If you use `ifOfferAllowed()` both of these filters can use method references instead. I find that's usually preferred when possible.

ghubstan replied:
Requested change in commit d709338d6366294ef727b0ab561e7ba73384872a .

sqrrm commented on desktop/src/main/java/bisq/desktop/main/offer/offerbook/OfferBook.java:
```diff
@@ -148,35 +148,52 @@ public void removeOffer(Offer offer, P2PDataStorage.ByteArray hashOfPayload, int
+            if (!isStoredLocally(offer)) {
+                Optional<OfferBookListItem> viewItem = getOfferBookListItem(offer);
+                viewItem.ifPresent((item) -> {
+                    offerBookListItems.remove(item);
```
Why is this needed? Isn't this just to present it nicely to the UI and that's not a problem when running the cli?

ghubstan replied:
>> Why is this needed?

Short explanation is a stale offer with no matching offer (with id) in local protected storage cannot be taken and should be removed from view.

>> Isn't this just to present it nicely to the UI and that's not a problem when running the cli?

The `OfferBookListItem` is a JFX observable used for display in the OfferBook views.  This code block supports the CLI being able to atomically edit and deactivate an offer in the same `editoffer` command, something the UI never does.  When you edit an offer with the UI, you deactivate the offer in the 1st mouse click (pencil button), then re-activate it with the confirm-edit or cancel button clicks -- two separate actions that ensure the temporarily deactivated offer is removed from peers, and the edited (or not) offer is re-published to peers (re-activated).

When the CLI edits _and_ deactivates an offer in the same `editoffer` command, the edited&disabled offer _will not be re-published_ by the API.  On another UI/Peer, the old stale OfferBookListItem would remain (without this code block) in the UI's observable item list because the peer receives on `onRemoved` msg containing the payload-hash of the newly edited offer the peer never received;  there is no matching payload-hash, and `OfferBook.onRemoved(offer, payloadHash)` does not know which item to remove.

ghubstan replied:
Commit fb4e00fb6bf6c6fbac536292d3b930f3c04cacf0 fixes the bug that caused me to add this workaround, and it has been removed in PR https://github.com/bisq-network/bisq/pull/5666

ghubstan committed:
Refactor 2 predicates as single predicate

Resolves https://github.com/bisq-network/bisq/pull/5659#discussion_r689631147

ghubstan committed:
Fix long method name

Resolves https://github.com/bisq-network/bisq/pull/5659#discussion_r687839333

ghubstan commented:
File conflicts resolved in child PR https://github.com/bisq-network/bisq/pull/5666
